src/core/thread/anycast_locator.hpp

```diff
@@ -0,0 +1,114 @@
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file includes definitions for Anycast Locator functionality.
+ */
+
+#ifndef ANYCAST_LOCATOR_HPP_
+#define ANYCAST_LOCATOR_HPP_
+
+#include "openthread-core-config.h"
+
+#if OPENTHREAD_CONFIG_TMF_ANYCAST_LOCATOR_ENABLE
+
+#include "coap/coap.hpp"
+#include "common/locator.hpp"
+#include "common/non_copyable.hpp"
+#include "net/ip6_address.hpp"
+
+namespace ot {
+
+/**
+ * This class implement Anycast Locator functionality which allows caller to determine the mesh local EID and RLOC16 of
+ * the closest destination of an anycast address (if any).
+ *
+ * The closest destination is determined based on the the current routing table and path costs within the Thread mesh.
+ *
+ */
+class AnycastLocator : public InstanceLocator, private NonCopyable
+{
+public:
+    /**
+     * This function pointer type defines the callback to notify the outcome of a request.
+     *
+     */
+    typedef otThreadAnycastLocatorCallback Callback;
+
+    /**
+     * This constructor initializes the `AnycastLocator` object.
+     *
+     * @param[in]  aInstance  A reference to the OpenThread instance.
+     *
+     */
+    explicit AnycastLocator(Instance &aInstance);
+
+    /**
+     * This method requests the closest destination of a given anycast address to be located.
+     *
+     * If a previous `Locate()` request is ongoing, a subsequent call to this method will cancel and replace the
+     * earlier request.
+     *
+     * @param[in] aAnycastAddress   The anycast address to locate the closest destination of.
+     * @param[in] aCallback         The callback handler to report the result.
+     * @param[in] aContext          An arbitrary context used with @p aCallback.
+     *
+     * @retval kErrorNone         The request started successfully. @p aCallback will be invoked to report the result.
+     * @retval kErrorNoBufs       Out of buffer to prepare and send the request message.
+     * @retval kErrorInvalidArgs  The @p aAnycastAddress is not a valid anycast address or @p aCallback is `nullptr`.
```

Comment:
Taking the same example as in the README, before srp server is enabled, it would take 60+ s to return timeout for the CON CoAP transaction if locating a valid ALOC which does not belong to any node yet.

```
> locate fd11:269a:9cbf:892d:0:ff:fe00:fc10
> Locate result
Error 28: ResponseTimeout
```

In fact we can learn whether the valid ALOC is available from network data in the mesh, similar as the mesh forwarder component 
https://github.com/openthread/openthread/blob/e0a1987336de203cdae30f06ff16cc2309817387/src/core/thread/mesh_forwarder_ftd.cpp#L546-L586

Maybe we can do some optimization here to only initiate `a/yl` CON CoAP msg when we know there must be some device has this valid ALOC?

Reply:
I noticed the longer timeout when there is no response. `AnycastLocator` uses the default COAP tx paramters which does up to 4 retries with some long randomr etry wait time, so the overall retries can take longer (up to even minutes). We can consider using a different `Coap::TxParameters` to make the timeout shorter but I am not sure if there is any advantages for this.

Related to determining if the anycast address is available before sending the CoAP message, we can consider it but again I am not sure of its benefit and whether the the complexity and code size impact it can add is worth such an optimization:
- `AnycastLocator` module needs to run on all types of builds/devices (MTD and FTD), so it's good to keep it lean (smaller code size).
- The code in `mesh_forwarder_ftd` is under FTD only and I think it requires that device to know the full network data (which may not be the case on an MTD build).

Reply:
> I noticed the longer timeout when there is no response. `AnycastLocator` uses the default COAP tx paramters which does up to 4 retries with some long randomr etry wait time, so the overall retries can take longer (up to even minutes). We can consider using a different `Coap::TxParameters` to make the timeout shorter but I am not sure if there is any advantages for this.

I think we should use the default coap to parameters.

> - `AnycastLocator` module needs to run on all types of builds/devices (MTD and FTD), so it's good to keep it lean (smaller code size).
> 
> - The code in `mesh_forwarder_ftd` is under FTD only and I think it requires that device to know the full network data (which may not be the case on an MTD build).

Checking network data for FTF only seems to be a good optimization. But I am wondering how often a FTD needs to use AnycastLocator.

Reply:
Just think of some questions - how does the caller know the ALOC it want to locate?
Should the general use case is the caller monitors the network data event and constructs the ALOC address when there are services (service id, context id, or commissioner session id etc) it cares and then `locate`? 

If we have such assumption, I think we should leave it as is; otherwise maybe we can consider this check optimization for Thread Devices (FTD or MTD) that requests full network data, avoiding unnecessary CoAP transaction and retx (which occupies limited message buffer resources for a while) and return result sooner, with the cost of iterating the network data once (?) (since we can know what services the ALOC is according to the address construction rule).

Reply:
If an FTD with full Network Data sends `a/yl` for an invalid ALOC, will it fail immediately because no route can be found?
If so, there won't be a long CoAP transaction with multiple retx, anyway. 

Another thought is that when a `a/yl` message is sent from an MTD child to its parent for an invalid ALOC, then the parent can't find any route. Will it send an `ICMPv6 Destination Unreachable` back to MTD? And will the MTD stop CoAP retx receiving `ICMPv6 Destination Unreachable` ?

Reply:
> Should the general use case is the caller monitors the network data event and constructs the ALOC address when there are services (service id, context id, or commissioner session id etc) it cares and then locate? 
>
> If we have such assumption, I think we should leave it as is; 

I agree. I think this is the intended use-case and assumption for this. The main use-case for now for this feature is DNS push related where we discover the anycast address itself from the network-data  (from the "DNS/SRP Service" entry in in network data) and then want to locate the destination of the anycast address.

> Will it send an ICMPv6 Destination Unreachable back to MTD? And will the MTD stop CoAP retx receiving ICMPv6 Destination Unreachable ?

This is an interesting idea, but I think we may want to consider it as a behavior to add in `Coap` itself (i.e., if we get an ICMPv6 "des unreachable" error we stop a CoAP transaction (not perform retx anymore). I think we can leave it for a future enhancement and PR.

Reply:
> If an FTD with full Network Data sends `a/yl` for an invalid ALOC, will it fail immediately because no route can be found?
> If so, there won't be a long CoAP transaction with multiple retx, anyway.

The upper layer CoAP retx is independent from no route decision at mesh forwarder layer, that is,  CoAP layer only knows there is no response (either due to not sent, or not arrived etc), then retx.




> > Will it send an ICMPv6 Destination Unreachable back to MTD? And will the MTD stop CoAP retx receiving ICMPv6 Destination Unreachable ?
> 
> This is an interesting idea, but I think we may want to consider it as a behavior to add in `Coap` itself (i.e., if we get an ICMPv6 "des unreachable" error we stop a CoAP transaction (not perform retx anymore). I think we can leave it for a future enhancement and PR.

Agree 👍